File: coordinator-events-backend/src/main/kotlin/hu/attilakillin/coordinatoreventsbackend/dal/Event.kt
package hu.attilakillin.coordinatoreventsbackend.dal

import jakarta.persistence.*
import java.time.OffsetDateTime

/**
 * The main domain specific entity of the application:
 * An event, with a title, and a list of emails registered to it.
 */
@Entity(name = "events")
data class Event(
    /** Unique ID, used internally, not shown to the end user. */
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,

    /** The moment an event was created. */
    var created: OffsetDateTime,

    /** The title of the event. */
    var title: String,

    /** The list of registered participant emails to this event, and their check-in status */
    @ElementCollection(fetch = FetchType.EAGER)
    @MapKeyColumn(name = "email")
    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    @CollectionTable(name = "participants", joinColumns = [JoinColumn(name = "id")])
    var participants: MutableMap<String, CheckinStatus>
)

File: coordinator-frontend/src/components/types/event.ts
export interface EventSummary {
    /** The unique ID of the event. Use this to request the whole content of the event. */
    id: number,

    /** The moment an event was created. */
    created: string,

    /** The title of the event. */
    title: string,
    /** The number of participants already registered to the event. */
    participants: number
};

export interface EventDetails {
    /** The unique ID of the event. Use this to request the whole content of the event. */
    id: number,

    /** The moment an event was created. */
    created: string,

    /** The title of the event. */
    title: string,
    /** The list of participants already registered to the event. */
    participants: string[]
};
